content/advent-2017/osquery-sdk.md

```diff
@@ -0,0 +1,274 @@
++++
+date = "2017-12-17T06:40:42Z"
+linktitle = "Extending Osquery with Go"
+title = "Extending Osquery with Go"
+author = ["Victor Vrantchan"]
++++
+
+What if you could use SQL to query any aspect of your infrastructure? [Osquery](https://github.com/facebook/osquery#osquery), an open source instrumentation tool released by the Facebook security team allows you to do just that. 
+
+For example, `SELECT network_name, last_connected, captive_portal FROM wifi_networks WHERE captive_portal=1;` will show all captive portal WiFi networks that a laptop has connected to. 
+And `SELECT * FROM processes WHERE on_disk = 0;` will show any process that is running where the binary has been deleted from disk.  
+When the root password vulnerability became know a few weeks ago, the osquery community [quickly crafted a query](https://twitter.com/osquery/status/935688822678892545) which would identify vulnerable macs in a fleet of devices. 
+With almost 200 tables available by default and supporting mac, linux and windows hosts, osquery is the tool of choice for many security and system administration teams. 
+
+Osquery is a powerful tool, but it’s written in C++, so why are we talking about it in a GopherAcademy post? Osquery uses Thrift(a project similar to gRPC) to allow developers to extend osquery through a series of plugin types. Earlier this year our team at [Kolide](https://kolide.com/) released a set of [Go packages](https://github.com/kolide/osquery-go) with idiomatic interfaces that allow anyone to use the full power of Go to extend osquery. In this blog post, it’s my goal to show you how you can get started with osquery development using the `osquery-go` SDK. 
+
+
+## Writing a custom logger plugin 
+
+When schedule a query like `SELECT name, version from deb_packages;` the `oqueryd` daemon will create a JSON log event with the results of the query. By default, a `filesystem` plugin is used, which logs the results to a local file. Commonly oquery users use aggregation tools like `filebeat`to send the result logs to a centralized log platform. Other plugins exist too. The `tls` plugin sends all logs to a remote TLS server like [Fleet](https://kolide.com/fleet). The `kinesis` plugin sends logs results to AWS, allowing advanced monitoring with applications like [StreamAlert](https://medium.com/airbnb-engineering/streamalert-real-time-data-analysis-and-alerting-e8619e3e5043). But what if you already have a well established logging pipeline with the systemd `journa`l, Splunk, `fluentd` or any number of proprietary logging systems. With the Thrift bindings to osquery, you can write your own logger. Go, having support for most APIs these days, is an ideal language for implementing a logger. 
```

Comment:
When scheduling*

Reply:
`journa`l should be `journal`

Reply:
Should there be a space between "`filebeat`" and "to"?